### PR TITLE
[lldb] Rewrite Swift completion logic in REPL to not crash [WIP]

### PR DESCRIPTION
**Pull request by Teemperor**
This fixes the crashes in the Swift REPL.

Fixes rdar://58854651

### CODE REVIEW

File: lldb/include/lldb/Utility/CompletionRequest.h [hunk 1]
```diff
@@ -104,7 +104,7 @@ class CompletionRequest {
 
   llvm::StringRef GetRawLine() const { return m_command; }
   llvm::StringRef GetRawLineUntilCursor() const {
-    return m_command.substr(0, m_cursor_index);
+    return m_command.substr(0, m_raw_cursor_pos);
```
Comment (dcci) [inline lldb/include/lldb/Utility/CompletionRequest.h:107]: Do we have a way to test this change? Is it tested as part of your other commits?

**Comment by Teemperor** [comment]
@swift-ci test

File: lldb/include/lldb/Utility/CompletionRequest.h (hunk 1 above)
Comment (Teemperor) [reply lldb/include/lldb/Utility/CompletionRequest.h:107]: Two of the commits here are not Swift-specific so I'll make a separate PR for them to the stable branch once this passes the test.

Comment (Teemperor) [reply lldb/include/lldb/Utility/CompletionRequest.h:107]: Whoops, somehow my commented ended up in this thread.... I just pushed a unit test for this change too.

**Comment by Teemperor** [comment]
@swift-ci test

**Comment by Teemperor** [comment]
Seems like the REPL fails to start as it doesn't reach the start prompt `>1` and just gets a EOF. The test works fine locally so I wonder what's going on there...

**Comment by Teemperor** [comment]
@swift-ci test

**Comment by Teemperor** [comment]
@swift-ci test

**Comment by Teemperor** [comment]
@swift-ci test

**Review by dcci — APPROVED** [review]
This looks great. Thanks for taking care of it.

**Comment by dcci** [comment]
[We should try to get this to 4.2]

**Comment by dcci** [comment]
I mean, 5.2

**Comment by Teemperor** [comment]
Split up in #678 and #679 (and made that against 5.2).

**Comment by marcrasi** [comment]
Do you still plan to merge this PR into `swift/master`? I'm asking because my REPL built at `swift/master` crashes when completing things and I wonder if this will fix it. e.g.:
```
marcrasi@marcrasi:~/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64$ bin/lldb -r
Welcome to Swift version 5.2-dev (LLVM 3bf86cf22b, Swift f4cef5715a).
Type :help for assistance.
  1> prin<TAB>
```
causes:
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::substr: __pos (which is 4) > this->size() (which is 2)
 #0 0x0000000000421ec4 PrintStackTraceSignalHandler(void*) (bin/lldb+0x421ec4)
 #1 0x000000000041ff5e llvm::sys::RunSignalHandlers() (bin/lldb+0x41ff5e)
 #2 0x000000000042244c SignalHandler(int) (bin/lldb+0x42244c)
 #3 0x00007fe2113e1520 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x13520)
 #4 0x00007fe208727081 raise (/lib/x86_64-linux-gnu/libc.so.6+0x3a081)
 #5 0x00007fe208712535 abort (/lib/x86_64-linux-gnu/libc.so.6+0x25535)
 #6 0x00007fe208961643 (/lib/x86_64-linux-gnu/libstdc++.so.6+0x9a643)
 #7 0x00007fe20896d066 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa6066)
 #8 0x00007fe20896d0b1 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa60b1)
 #9 0x00007fe20896d2e5 (/lib/x86_64-linux-gnu/libstdc++.so.6+0xa62e5)
#10 0x00007fe208963e0d (/lib/x86_64-linux-gnu/libstdc++.so.6+0x9ce0d)
#11 0x00007fe209972329 lldb_private::Editline::TabCommand(int) (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xd2e329)
#12 0x00007fe209975039 lldb_private::Editline::ConfigureEditor(bool)::$_14::__invoke(editline*, int) (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xd31039)
#13 0x00007fe2082fe792 el_wgets (/lib/x86_64-linux-gnu/libedit.so.2+0x10792)
#14 0x00007fe209974842 lldb_private::Editline::GetLines(int, lldb_private::StringList&, bool&) (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xd30842)
#15 0x00007fe2098e7324 lldb_private::IOHandlerEditline::Run() (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xca3324)
#16 0x00007fe2098cbb56 lldb_private::Debugger::ExecuteIOHandlers() (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xc87b56)
#17 0x00007fe2098cfb7f lldb_private::Debugger::IOHandlerThread(void*) (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xc8bb7f)
#18 0x00007fe20997e062 lldb_private::HostNativeThreadBase::ThreadCreateTrampoline(void*) (/usr/local/google/home/marcrasi/swift-base-master/build/Ninja-ReleaseAssert/lldb-linux-x86_64/bin/../lib/liblldb.so.7git+0xd3a062)
#19 0x00007fe2113d6fb7 start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8fb7)
#20 0x00007fe2087e72df clone (/lib/x86_64-linux-gnu/libc.so.6+0xfa2df)
********************
```

**edit**: Indeed, I patched this in and it fixes that crash.

**Comment by Teemperor** [comment]
Ah, yeah, this patch landed in 5.2, I need to do the same PR for all the other branches I guess. Thanks for the hint

**Comment by Teemperor** [comment]
@marcrasi it's on master now, thanks!

**Pull request closed without merging**